### PR TITLE
Add dual stack network tests

### PR DESCRIPTION
**Pull request by martencassel**
fixes https://github.com/k0sproject/k0s/issues/2653

Signed-off-by: Mårten Cassel <marten.cassel@gmail.com>

### CODE REVIEW

**Comment by martencassel** [comment]
I cant see why this fails in pipeline.

**Comment by jnummelin** [comment]
```
    footloosesuite.go:199: 
        	Error Trace:	/home/runner/work/k0s/k0s/inttest/dualstack/footloosesuite.go:199
        	            				/home/runner/work/k0s/k0s/inttest/dualstack/asm_amd64.s:1598
        	Error:      	Test deadline exceeded
        	Test:       	TestDualStack
    footloosesuite.go:202: Cleaning up
    footloosesuite.go:209: Test cleanup deadline: 2023-02-08 23:07:36.547129158 +0000 UTC m=+220.012969228
    dualstack_test.go:130: 
        	Error Trace:	/home/runner/work/k0s/k0s/inttest/dualstack/dualstack_test.go:130
        	            				/home/runner/work/k0s/k0s/inttest/dualstack/suite.go:132
        	            				/home/runner/work/k0s/k0s/inttest/dualstack/dualstack_test.go:241
        	Error:      	Received unexpected error:
        	            	context deadline exceeded
        	Test:       	TestDualStack
        	Messages:   	nginx pod did not start
```
For some reason the smoke was not able to gather all the logs 🤔 

Do you have the possibility to run it locally?

**Comment by martencassel** [comment]
I have and see no such errors

[dualstack.log](https://github.com/k0sproject/k0s/files/10695940/dualstack.log)

File: inttest/common/footloosesuite.go [hunk 1]
```diff
@@ -1444,3 +1445,37 @@ func (s *FootlooseSuite) AssertSomeKubeSystemPods(client *kubernetes.Clientset)
...
+	}
+	err = ioutil.WriteFile(tmpFile.Name(), json, 0644)
+	if err != nil {
+		return fmt.Errorf("failed to write to temporary file: %w", err)
+	}
+	err = exec.Command("sudo", "mv", "-f", tmpFile.Name(), "/etc/docker/daemon.json").Run()
+	if err != nil {
+		return fmt.Errorf("failed to write daemon.json file: %w", err)
+	}
+	err = exec.Command("sudo", "systemctl", "reload", "docker").Run()
```
Comment (jnummelin) [inline inttest/common/footloosesuite.go:1476]:
```suggestion
	err = exec.Command("sudo", "systemctl", "restart", "docker").Run()
```

IIRC `reload` just reloads the unit from disk but does NOT actually restart it

Comment (martencassel) [reply inttest/common/footloosesuite.go:1476]: sure we can try that
i folloed https://docs.docker.com/config/daemon/ipv6/

**Comment by jnummelin** [comment]
I think there's one fundamental _thing_ we need to decide on. So the current version in this PR changes the docker config via code and restarts the unit. That has couple general challenges:
- tied to systemd; remember many people use e.g. osx as the dev env
- docker config file location hard coded

I'm afraid that there's too much variation in the envs (linux vs. osx vs. all the possible Linux distros) where we run the smokes to handle docker config at the test code.

What I'm instead of thinking is if we could detect whether docker has ipv6 support enabled in the test case. If it's enabled, then run the test case as expected. If it's NOT enabled, bail out with some generic help info for the user how to enable it in docker. Then in smokes we could enable it via actions.

So essentially we "externalize" the docker ipv6 config for the user and/or to CI setup.

@twz123 You've had some thoughts on this too, any comments?

File: inttest/common/footloosesuite.go [hunk 2]
```diff
@@ -1444,3 +1445,37 @@ func (s *FootlooseSuite) AssertSomeKubeSystemPods(client *kubernetes.Clientset)
 
 	return false
 }
+
+type IPV6Config struct {
+	IPv6        bool   `json:"ipv6"`
+	FixedCidrV6 string `json:"fixed-cidr-v6"`
+}
+
+func (s *FootlooseSuite) ConfigureIPv6Docker(enabled bool) error {
```
Comment (twz123) [inline inttest/common/footloosesuite.go:1454]: This is something that's host-specific. Better put it into the GitHub workflows? Maybe as a shell script that's invoked as part of a `run` script?

Comment (martencassel) [reply inttest/common/footloosesuite.go:1454]: That sounds as a good idea i can look into it.
restarting the docker daemon must also be done before the footloose suite creates all containers.

Comment (martencassel) [reply inttest/common/footloosesuite.go:1454]: One way to test if IPv6 is enabled could be to check if the default docker bridge network has EnableIPv6 set to true.

Comment (martencassel) [reply inttest/common/footloosesuite.go:1454]: Should we bail out with an error or not ?

**Comment by martencassel** [comment]
I added a test for IPv6 config in docker, if not enabled then fail.

File: inttest/dualstack/dualstack_test.go [hunk 3]
```diff
@@ -94,9 +111,122 @@ func (s *DualstackSuite) SetupSuite() {
...
+		}
+		s.T().Log("server response", out)
+		return strings.Contains(out, "Welcome to nginx"), nil
+	})
+	s.Require().NoError(err)
+
 	s.client = client
 }
 
+func getAlpineVersion(t *testing.T) string {
```
Comment (juanluisvaladas) [inline inttest/dualstack/dualstack_test.go:220]: You can remove all this logic if you use nginx for both pods. The nginx image has both wget and curl so you can remove this.

**Review by juanluisvaladas — COMMENTED** [review]
The test in general looks quite good but it can be made faster by using nginx pods as clients.

You can create 2 ngnix pods called nginx-worker0 and nginx-worker1 and then have nginx-worker-0 curl/wget both nginx-worker-1 ip addresses. You are effectively testing the same things but you have to pull half the images and run half the containers.

Also this allows to reuse more code and would make it cleaner.

Other than that, the test looks good to me, now that docker isn't being reconfigured I don't see any red flags and the test is useful and tests what it has to test.

File: inttest/dualstack/dualstack_test.go (hunk 3 above)
Comment (martencassel) [reply inttest/dualstack/dualstack_test.go]: I did as you suggested, see new commit.

**Review by twz123 — CHANGES_REQUESTED** [review]
Thanks for working on this @martencassel. "Works on my machine." :+1: To make the integration tests pass on the GitHub runners, we need to enable IPv6 in Docker on those. That's basically the things you did [here][enable-ipv6]. We just need them as another build step in the `smoketest` job in `.github/workflows/go.yml`, somewhere after (or as part of) the `Prepare build environment` step. Would you mind to rewrite that as a shell script and add add it there?

Moreover, there are some lints concerning variable naming. We're not currently linting the integration tests during the lint CI job, nevertheless it would be great if you could address those:

```
inttest/dualstack/dualstack_test.go:147:3: var-naming: don't use underscores in Go names; var pod_ip should be podIP (revive)
                pod_ip := targetPod.Status.PodIPs[1].IP
                ^
inttest/dualstack/dualstack_test.go:148:3: var-naming: don't use underscores in Go names; var target_ip should be targetIP (revive)
                target_ip := net.ParseIP(pod_ip)
                ^
inttest/dualstack/dualstack_test.go:163:3: var-naming: don't use underscores in Go names; var pod_ip should be podIP (revive)
                pod_ip := targetPod.Status.PodIPs[0].IP
                ^
inttest/dualstack/dualstack_test.go:164:3: var-naming: don't use underscores in Go names; var target_ip should be targetIP (revive)
                target_ip := net.ParseIP(pod_ip)
                ^
inttest/common/footloosesuite.go:1454:6: var-naming: don't use underscores in Go names; var docker_bridge_EnableIPv6 should be dockerBridgeEnableIPv6 (revive)
        var docker_bridge_EnableIPv6 string
            ^
inttest/common/footloosesuite.go:1459:2: var-naming: don't use underscores in Go names; var bridge_EnableIPv6 should be bridgeEnableIPv6 (revive)
        bridge_EnableIPv6, err := strconv.ParseBool(docker_bridge_EnableIPv6)
        ^
```

[enable-ipv6]: https://github.com/k0sproject/k0s/commit/9ae95fdac6dbf8cc598734a256c5919d1f84dd66#diff-85e4a1b54e18b85233536b2d526e695deea45e09bd65ecf7e4ed55a7c1441b4eR1454-R1481

File: inttest/dualstack/dualstack_test.go (hunk 3 above)
Comment (juanluisvaladas) [reply inttest/dualstack/dualstack_test.go]: Looks better, we need to do change some stuff on the CI before merging. Other than the linting stuff that Tom mentioned it looks good

**Review by twz123 — APPROVED** [review]

**Comment by twz123** [comment]
Ha! Enabling IPv6 in Docker has side effects on the airgap integration test :tada:

I'll have a look into that.

**Comment by twz123** [comment]
You could add a condition to the "Prepare docker for ipv6 dualstack tests" build step, like this: `if: contains(matrix.smoke-suite, 'dualstack')`.

That way, we can proceed here and can fix the airgap test in subsequent PRs.

**Review by twz123 — APPROVED** [review]

**Comment by twz123** [comment]
Thanks again!

**Comment by martencassel** [comment]
Thankyou it was pure fun :)

**Pull request merged**